src/editor.rs

```diff
@@ -353,58 +344,75 @@ impl<'arena, Node: Ast<'arena> + 'arena> Editor<'arena, Node> {
-                    log::debug!("Inserting with '{}'", c);
-                }
-            } else {
+        //short circuit if not an insertable char
+        if !cursor.is_insert_char(c) {
+            log::warn!("Cannot insert node with '{}'", c);
+            return;
+        }
+
+        //short circuit if char is invalid
+        let node = match cursor.from_char(c) {
+            Some(node) => node,
+            None => {
                 log::warn!("Char '{}' does not correspond to a valid node", c);
+                return;
             }
-        } else {
-            log::warn!("Cannot insert node with '{}'", c);
+        };
+
+        match self.tree.insert_child(node) {
+            Ok(_) => log::debug!("Inserting with '{}'", c),
+            Err(e) => log::error!("{}", e),
         }
     }
 
     /// Insert new child as the first child of the selected node
     fn insert_next_to_cursor(&mut self, c: char, side: Side) {
-        let (_cursor, parent) = self.tree.cursor_and_parent();
-        if let Some(p) = parent {
-            if p.is_insert_char(c) {
-                if let Some(node) = p.from_char(c) {
-                    if let Err(e) = self.tree.insert_next_to_cursor(node, side) {
-                        log::error!("{}", e);
-                    } else {
-                        log::debug!("Inserting with '{}'", c);
-                    }
-                } else {
-                    log::warn!("Char '{}' does not correspond to a valid node", c);
-                }
-            } else {
-                log::warn!("Cannot insert node with '{}'", c);
+        let (_, parent) = self.tree.cursor_and_parent();
+
+        //Short circuit if at root
+        let parent = match parent {
+            Some(parent) => parent,
+            None => {
+                log::warn!("Cannot add siblings of the root.");
+                return;
+            }
+        };
+
+        //short circuit if not an insertable char
+        if !parent.is_insert_char(c) {
+            log::warn!("Cannot insert node with '{}'", c);
+            return;
+        }
+
+        //short circuit if invalid char
+        let node = match parent.from_char(c) {
+            Some(node) => node,
+            None => {
+                log::warn!("Char '{}' does not correspond to a valid node", c);
+                return;
             }
+        };
+
+        if let Err(e) = self.tree.insert_next_to_cursor(node, side) {
+            log::error!("{}", e);
         } else {
-            log::warn!("Cannot add siblings of the root.");
+            log::debug!("Inserting with '{}'", c);
         }
     }
 
     /// Undo the latest change
     fn undo(&mut self) {
-        if self.tree.undo() {
-            log::debug!("Undo successful");
-        } else {
-            log::warn!("No changes to undo");
+        match self.tree.undo() {
+            true => log::debug!("Undo successful"),
+            false => log::warn!("No changes to undo"),
         }
```

Comment:
I'm unconvinced that this is easier to read than a simple `if`/`else`, but I have an idea to move the error-checking code down into `DAG` anyway, so it's probably no big deal.

@Kixunil, what do you reckon?

Reply:
Tbh I reckon that `self.tree.undo()`/`Dag::undo()` should return a `Result<(), ...>` anyway, but that's refactoring for another day.

Reply:
I occasionally write `match` like that as well, but I'm unable to explain why/when. Just intuition.
It's a bit shorter but perhaps surprising?

Reply:
To be honest, I've warmed to the idea... because it is essentially doing error handling, I think it feels better to use a `match`.